parmys-plugin/parmys.cc

```diff
@@ -0,0 +1,1175 @@
+/*
+ * Copyright 2022 Daniel Khadivi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "kernel/celltypes.h"
+#include "kernel/yosys.h"
+
+#include <regex>
+
+#include "netlist_utils.h"
+#include "odin_globals.h"
+#include "odin_ii.h"
+#include "odin_util.h"
+
+#include "vtr_memory.h"
+#include "vtr_path.h"
+#include "vtr_util.h"
+
+#include "netlist_check.h"
+
+#include "partial_map.h"
+
+#include "netlist_visualizer.h"
+
+#include "parmys_resolve.hpp"
+
+#include "BlockMemories.hpp"
+#include "adders.h"
+#include "arch_util.h"
+#include "hard_blocks.h"
+#include "memories.h"
+#include "multipliers.h"
+#include "netlist_cleanup.h"
+#include "netlist_statistic.h"
+#include "netlist_visualizer.h"
+#include "read_xml_config_file.h"
+#include "subtractions.h"
+
+#include "ast_util.h"
+#include "parmys_update.hpp"
+#include "parmys_utils.hpp"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+#define GND_NAME "$false"
+#define VCC_NAME "$true"
+#define HBPAD_NAME "$undef"
+
+struct Bbox {
+    std::string name;
+    std::vector<std::string> inputs, outputs;
+};
+
+CellTypes ct;
+
+struct ParMYSPass : public Pass {
+
+    static void hook_up_nets(netlist_t *odin_netlist, Hashtable *output_nets_hash)
+    {
+        nnode_t **node_sets[] = {odin_netlist->internal_nodes, odin_netlist->ff_nodes, odin_netlist->top_output_nodes};
+        int counts[] = {odin_netlist->num_internal_nodes, odin_netlist->num_ff_nodes, odin_netlist->num_top_output_nodes};
+        int num_sets = 3;
+
+        int i;
+        for (i = 0; i < num_sets; i++) {
+            int j;
+            for (j = 0; j < counts[i]; j++) {
+                nnode_t *node = node_sets[i][j];
+                hook_up_node(node, output_nets_hash);
+            }
+        }
+    }
+
+    static void hook_up_node(nnode_t *node, Hashtable *output_nets_hash)
+    {
+        int j;
+        for (j = 0; j < node->num_input_pins; j++) {
+            npin_t *input_pin = node->input_pins[j];
+
+            nnet_t *output_net = (nnet_t *)output_nets_hash->get(input_pin->name);
+
+            if (!output_net)
+                log_error("Error: Could not hook up the pin %s: not available, related node: %s.", input_pin->name, node->name);
+            add_fanout_pin_to_net(output_net, input_pin);
```

Comment:
this will be called even if `output_net` is NULL (which you check above). Shouldn't this function simply return on the error (and maybe return some error value so caller can react(?

Reply:
AFAIK the `log_error` function of Yosys throws an exception while logging the error. 
And here we expect the runtime to be interrupted if `output_net` is NULL.
You mean if `output_net` is NULL the `add_fanout_pin_to_net` is still called?